### PR TITLE
restore: remove tiflash replica before restore

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

### What problem does this PR solve? 
Support restoring to a cluster containing TiFlash, see https://github.com/pingcap/br/issues/193

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change

Side effects

 - Possible performance regression
 - Increased code complexity

### CODE REVIEW

File: pkg/restore/db.go [hunk 1]
```diff
@@ -129,6 +129,21 @@ func (db *DB) CreateTable(ctx context.Context, table *utils.Table) error {
 			zap.Stringer("table", table.Info.Name),
 			zap.Error(err))
 	}
+
+	// TODO: remove this after tiflash supports restore
+	removeTiFlashSQL := fmt.Sprintf(
+		"alter table %s set tiflash replica 0",
+		utils.EncloseName(tableInfo.Name.O),
+	)
+	err = db.se.Execute(ctx, removeTiFlashSQL)
```
Comment (3pointer) [inline pkg/restore/db.go]: what if ignore this error when tidb not support set tiflash replica

**Comment by overvenus** [comment]
Should change `GetAllTiKVStore` on restore, it fails if there is a tiflash node.

File: pkg/restore/db.go (hunk 1 above)
Comment (5kbpers) [reply pkg/restore/db.go]: Would fail, do we need to support the versions of tidb which don't support tiflash?

**Comment by 3pointer** [comment]
Do we need add replica back after restore?

File: pkg/restore/client.go [hunk 2]
```diff
@@ -305,6 +305,71 @@ func (rc *Client) CreateTables(
...
+		err = rc.db.se.Execute(rc.ctx, removeTiFlashSQL)
+		if err != nil {
+			log.Error("remove tiflash replica failed",
+				zap.String("query", removeTiFlashSQL),
+				zap.Stringer("db", table.Db.Name),
+				zap.Stringer("table", table.Info.Name),
+				zap.Error(err))
+			return err
+		} else if replicaCount > 0 {
+			log.Warn("remove tiflash replica done, please recover it manually later",
```
Comment (3pointer) [inline pkg/restore/client.go]: what if print one executable DDL `alter table xxx set tiflash replica replicaCount`?

File: pkg/restore/client.go [hunk 3]
```diff
@@ -305,6 +305,71 @@ func (rc *Client) CreateTables(
...
+				zap.Stringer("db", table.Db.Name),
+				zap.Stringer("table", table.Info.Name),
+				zap.Int("originalReplicaCount", replicaCount))
+		}
+	}
+	return nil
+}
+
+// GetTiFlashStores returns an id list of tiflash stores.
+func (rc *Client) GetTiFlashStores() ([]uint64, error) {
```
Comment (overvenus) [inline pkg/restore/client.go]: Could you add an option to `GatAllTiKVStores` to return TiFlash store only?

File: pkg/restore/split.go [hunk 4]
```diff
@@ -192,6 +215,30 @@ func (rs *RegionSplitter) isScatterRegionFinished(ctx context.Context, regionID
...
+	}
+	for _, peer := range regionInfo.Region.GetPeers() {
+		if rejectStores[peer.GetStoreId()] {
+			return true, nil
+		}
+	}
+	retryTimes := ctx.Value(retryTimes).(int)
+	if retryTimes > 10 {
+		log.Warn("get region info", zap.Stringer("region", regionInfo.Region))
+	}
```
Comment (overvenus) [inline pkg/restore/split.go]: Could you handle the retry outside of the method?

File: pkg/restore/client.go [hunk 5]
```diff
@@ -305,6 +305,71 @@ func (rc *Client) CreateTables(
 	return rewriteRules, newTables, nil
 }
 
+// RemoveTiFlashReplica removes all the tiflash replica of a table
+func (rc *Client) RemoveTiFlashReplica(tables []*utils.Table) error {
+	stores, err := rc.GetTiFlashStores()
+	if err != nil {
+		return errors.Trace(err)
+	}
```
Comment (3pointer) [inline pkg/restore/client.go]: if len(stores) > 0  need return immediately

File: go.mod [hunk 6]
```diff
@@ -20,9 +20,10 @@ require (
 	github.com/onsi/gomega v1.8.1 // indirect
 	github.com/pingcap/check v0.0.0-20200212061837-5e12011dc712
 	github.com/pingcap/errors v0.11.5-0.20190809092503-95897b64e011
-	github.com/pingcap/kvproto v0.0.0-20200228095611-2cf9a243b8d5
+	github.com/pingcap/kvproto v0.0.0-20200317112120-78042b285b75
 	github.com/pingcap/log v0.0.0-20200117041106-d28c14d3b1cd
 	github.com/pingcap/parser v0.0.0-20200305120128-bde9faa0df84
+	github.com/pingcap/pd v2.1.19+incompatible
```
Comment (3pointer) [inline go.mod]: should remove this?

File: pkg/restore/client.go [hunk 7]
```diff
@@ -305,6 +334,87 @@ func (rc *Client) CreateTables(
...
+	backendURL := storage.FormatBackendURL(rc.backend)
+	log.Info("update backup meta", zap.Stringer("path", &backendURL))
+	err = rc.storage.Write(rc.ctx, utils.MetaFile, backupMetaData)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, table := range tables {
```
Comment (3pointer) [inline pkg/restore/client.go]: should we skip table doesn't need set replica?

File: pkg/restore/split.go (hunk 4 above)
Comment (5kbpers) [reply pkg/restore/split.go]: No, the region info would be printed here.

**Comment by 3pointer** [comment]
/run-integration-test pd=release-4.0 tidb=release-4.0 tikv=release-4.0

File: pkg/conn/conn.go [hunk 8]
```diff
@@ -103,6 +103,9 @@ const (
 	// SkipTiFlash causes GetAllTiKVStores to skip the store when it is found to
 	// be a TiFlash node.
 	SkipTiFlash UnexpectedStoreBehavior = 1
+	// TiFlashOnly caused GetAllTiKVStores to skip the store which is not a
+	// TiFlash node.
+	TiFlashOnly UnexpectedStoreBehavior = 2
```
Comment (overvenus) [inline pkg/conn/conn.go]: Could you rename `UnexpectedStoreBehavior` to `StoreBehavior`?

File: pkg/restore/client.go [hunk 9]
```diff
@@ -305,6 +334,70 @@ func (rc *Client) CreateTables(
...
+			Crc64Xor:        table.Crc64Xor,
+			TotalKvs:        table.TotalKvs,
+			TotalBytes:      table.TotalBytes,
+			TiflashReplicas: uint32(table.TiFlashReplicas),
+		})
+	}
+
+	// Update backup meta
+	rc.backupMeta.Schemas = schemas
+	backupMetaData, err := proto.Marshal(rc.backupMeta)
```
Comment (3pointer) [inline pkg/restore/client.go]: the origin backupmeta schemas has changed to new-created schemas, but it seems has no affect on next restore. will it affect on our current debug procedure?

File: cmd/restore.go [hunk 10]
```diff
@@ -30,6 +38,59 @@ func runRestoreRawCommand(command *cobra.Command, cmdName string) error {
...
+	defer store.Close()
+	db, err := restore.NewDB(tidbGlue, store)
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+
+	for _, d := range dbs {
+		for _, t := range d.Tables {
+			log.Info("get table", zap.Stringer("name", t.Info.Name), zap.Int("replica", t.TiFlashReplicas))
```
Comment (overvenus) [inline cmd/restore.go]: Could you `cmd.Println` the result?

File: pkg/restore/client.go [hunk 11]
```diff
@@ -305,6 +334,70 @@ func (rc *Client) CreateTables(
...
+	if err != nil {
+		return errors.Trace(err)
+	}
+	backendURL := storage.FormatBackendURL(rc.backend)
+	log.Info("update backup meta", zap.Stringer("path", &backendURL))
+	err = rc.storage.Write(rc.ctx, utils.MetaFile, backupMetaData)
```
Comment (overvenus) [inline pkg/restore/client.go]: Could you write to another file? Backup files are supposed to be immutable.

File: pkg/restore/client.go [hunk 12]
```diff
@@ -305,6 +334,70 @@ func (rc *Client) CreateTables(
...
+			Table:           tableData,
+			Crc64Xor:        table.Crc64Xor,
+			TotalKvs:        table.TotalKvs,
+			TotalBytes:      table.TotalBytes,
+			TiflashReplicas: uint32(table.TiFlashReplicas),
+		})
+	}
+
+	// Update backup meta
+	rc.backupMeta.Schemas = schemas
```
Comment (overvenus) [inline pkg/restore/client.go]: Can we skip write to storage if there is no TiFlash node?

File: pkg/utils/pd.go [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer resp.Body.Close()
+	buf := new(bytes.Buffer)
+	_, err = buf.ReadFrom(resp.Body)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if resp.StatusCode == 412 {
```
Comment (overvenus) [inline pkg/utils/pd.go]: 412 feels like magic.

File: cmd/restore.go (hunk 10 above)
Comment (5kbpers) [reply cmd/restore.go]: Added a progress bar.

File: pkg/task/restore.go [hunk 14]
```diff
@@ -349,3 +369,51 @@ func splitPostWork(ctx context.Context, client *restore.Client, tables []*model.
...
+	if err != nil {
+		return err
+	}
+
+	tables := make([]*utils.Table, 0)
+	for _, db := range dbs {
+		tables = append(tables, db.Tables...)
+	}
+	updateCh := utils.StartProgress(
+		ctx, "Checksum", int64(len(tables)), !cfg.LogProgress)
```
Comment (3pointer) [inline pkg/task/restore.go]:
```suggestion
		ctx, "RestoreTiflashReplica", int64(len(tables)), !cfg.LogProgress)
```

**Comment by 5kbpers** [comment]
/run-integration-test pd=release-4.0 tidb=release-4.0 tikv=release-4.0

**Pull request merged**